API refactor module to be more independent

NightJar opened the pull request:
As part of a greater piece of work with site summaries (cf.
https://github.com/bringyourownideas/silverstripe-maintenance/issues/39). This commit tidies class
definitions and makes the import slightly more efficient. Abstractions are
added to relate to the Package object from the aforementioned module.
Unit tests have been added to ensure tidy standards in future maintenance.

resolves #1 (although technically it was already resolved).

Commit:
API refactor module to be more independent

As part of a greater piece of work with site summaries (cf.
bringyourownideas/silverstripe-maintenance). This commit tidies class
definitions and makes the import slightly more efficient. Abstractions are
added to relate to the Package object from the aforementioned module.
Unit tests have been added to ensure tidy standards in future maintenance.

robbieaverill commented on composer.json:
```diff
@@ -19,15 +19,15 @@
         "spekulatius/silverstripe-composer-security-checker": "self.version"
     },
     "suggest": {
+        "friendsofsilverstripe/silverstripe-maintenance": "Display advisory information via Site Summary Report",
```
The package name has changed =)

robbieaverill commented on composer.json:
```diff
@@ -19,15 +19,15 @@
-        "sensiolabs/security-checker": "^3.0",
-        "symbiote/silverstripe-queuedjobs": "^2.8",
-        "friendsofsilverstripe/silverstripe-maintenance": "^0.3"
+        "silverstripe/framework": "^3",
```
Let's introduce `^3.2` which matches bringyourownideas/silverstripe-maintenance

robbieaverill commented on src/Jobs/CVECheckJob.php:
```diff
@@ -0,0 +1,61 @@
+    /**
+     * @return CVECheckTask
+     */
+    public function getCVECheckTask()
```
Can you please name the getters/setters the same as the property? I.e. `setCheckTask(CheckTask $checkTask)`

robbieaverill commented on src/Jobs/CVECheckJob.php:
```diff
@@ -0,0 +1,61 @@
+    }
+
+    /**
+     * @param CVECheckTask
```
Missing variable name - can you please type hint the class in the argument too?

robbieaverill commented on src/Tasks/CVECheckTask.php:
```diff
@@ -0,0 +1,118 @@
+     * @param SecurityChecker $securityChecker
+     * @return CVECheckTask $this
+     */
+    public function setSecurityChecker($securityChecker)
```
Please type hint SecurityChecker

robbieaverill commented on src/Tasks/CVECheckTask.php:
```diff
@@ -0,0 +1,118 @@
+
+    /**
+     * @param SecurityChecker $securityChecker
+     * @return CVECheckTask $this
```
Just `@return $this`

robbieaverill commented on src/Tasks/CVECheckTask.php:
```diff
@@ -0,0 +1,118 @@
+
+        // use the security checker of
+        $checker = $this->getSecurityChecker();
+        $alerts = $checker->check(BASE_PATH . DIRECTORY_SEPARATOR . 'composer.lock');
```
I feel like we should use the ComposerLoader from silverstripe-maintenance for something here, at the least to remove the reliance on the physical path of composer.lock from this code (and leave it to ComposerLoader)

NightJar replied:
1) This library has it's own logic for reading and parsing the lock file, and only requires the path in the check method.
2) ComposerLoader requires a base path, which means either specifying manually still, or relying on `BASE_PATH` as it is here.
3) There is no way at current to extract the lock path from ComposerLoader (only the parsed contents).
4) This modules does not require silverstripe-maintenance, nor is it intended to.
5) This task is for checking the main installation only, not another at some arbitrary location, nor does it provide the ability to upload and check random lock files.

**Conclusion:**
ComposerLoader should not be used for this. Although I could add a setter/getter for a lock path, defaulting to what is currently set, I feel it is unnecessary.

robbieaverill commented on src/Tasks/CVECheckTask.php:
```diff
@@ -0,0 +1,118 @@
+
+                // Is this vulnerability known? No, lets add it.
+                if ((int) $vulnerability->count() === 0) {
+                    $vulnerability = new CVE();
```
`CVE::create()`

robbieaverill commented on src/Tasks/CVECheckTask.php:
```diff
@@ -0,0 +1,118 @@
+        }
+
+        // remove all entries which are resolved (no longer $validEntries)
+        $removeOldCVEs = SQLDelete::create('CVE');
```
Please quote the table name to ensure that this works across DB drivers

robbieaverill commented on src/Tasks/CVECheckTask.php:
```diff
@@ -0,0 +1,118 @@
+            $removable = CVE::get()->exclude(array('ID' => $validEntries));
+            if ($removable->exists()) {
+                // Be careful not to remove all CVEs on the case that entry is valid
+                $removeOldCVEs = $removeOldCVEs->addWhere('ID', $removable->column('ID'));
```
Quotes for columns

robbieaverill commented on tests/CVECheckTaskTest.php:
`$this->assertCount(0, $preCheck, '...`

robbieaverill reviewed (changes requested):
Looks good overall! I've left some specific comments, but also there's generally a huge lack of PHPDocs on methods and classes

Commit:
Add more display output for security alerts

Show the information in summary to users looking at the SiteSummary report
on screen. This includes the ability to see a tag notifying which modules
have alerts issued for them, a form level warning alerting users that some
modules have alerts (as they may not display on the first page), and the
ability to expand an individual module's information to display the alerts
that have been issued (in a summary of title, ID number and externa link)

Commit:
Alter package level alert notice functionality

For screen rendering devices - as per design team feedback. The text
denoting an action is present is removed, and the toggle switch for
revealing that information is shifted to the badge itself.

NightJar commented:
Design review is done, this is ready for final review :)

robbieaverill commented on .editorconfig:
```diff
@@ -10,6 +10,13 @@ indent_style = space
 trim_trailing_whitespace = true
 
+[*.md]
+trim_trailing_whitespace = false
```
👎

NightJar replied:
😮

robbieaverill commented on src/Extensions/PackageSecruityExtension.php:
```diff
@@ -0,0 +1,57 @@
+    public function updateBadges(&$badges)
+    {
+        if ($this->owner->SecurityAlerts()->exists()) {
+            $badges['RISK: Security'] = 'warning security-alerts__toggler';
```
Can you please make this label translatable?

NightJar replied:
Oh whoops, yep :>

robbieaverill commented on templates/SecurityAlertSummary.ss:
```diff
@@ -0,0 +1,4 @@
+<p>
+    <strong>Security alert</strong><br />
+    <% if Count > 1 %>Notices have<% else %>A notice has<% end_if %> been issued for <strong>$Count</strong> of your modules. Review and updating is recommended.
```
Please add translations

robbieaverill reviewed (changes requested):
Looks great, nice work! I've left a couple of comments to add translations, but otherwise looks good

robbieaverill commented:
Can you add a branch alias (1.x-dev) to composer.json while you're at it?

Commit:
Make output strings translatable

Switch away from hard coded English.

robbieaverill commented on src/Extensions/PackageSecruityExtension.php:
```diff
@@ -0,0 +1,57 @@
+     *
+     * @param array $badges in the format of [title => type]
+     */
+    public function updateBadges(&$badges)
```
I'm hoping this will change in the other PR so will expect this to change

NightJar replied:
ooh, good call; derp.

robbieaverill replied:
Yep needs updating now to use an ArrayList

NightJar replied:
Is done :>

robbieaverill replied:
PHPDoc still needs updating

robbieaverill commented on templates/SecurityAlertSummary.ss:
```diff
@@ -0,0 +1,9 @@
+    <% if Count > 1 %>
+        <%t SecuirtyAlertSummary.NOTICES_HAVE "Notices have" %>
+    <% else %>
+        <%t SecuirtyAlertSummary.NOTICE_HAS "A notice has" %>
```
On reflection I don’t like how these translations are split up. It provides no context to translators and they might get confused about it. Also typos in the key :-)

robbieaverill commented on src/Models/CVE.php:
```diff
@@ -0,0 +1,20 @@
+/**
+ * Describes a known security issue of an installed Composer package
+ */
+class CVE extends DataObject
```
Last thing- the vulnerabilities are not always assigned CVEs (ours aren’t for example), so maybe the original name or something like PackageVulnerability would be better- what do you think?

NightJar replied:
Question: Should the SS security ID then be put in place of the CVE ID?
i.e. `SS-2015-014` when accessing `->CVE`?

robbieaverill replied:
No, because this module can work for any PHP packages not just SilverStripe packages. Maybe choose something generic like "Identifier"

Commit:
Big refactor to make sense of code

Use sensible names, and also fix a few bug discovered by altering the
tests with more data and cases (such as the SQLDelete statement). CVE
DataObject is now SecurityAlerts, as not all alerts are assigned CVE
identifiers. To that end if there is another identifier (such as the one
SilverStripe use as a title prefix) then this is set instead. As such the
property is renamed 'Identifier' as opposed to 'CVE'.

robbieaverill commented on src/Jobs/SecurityAlertCheckJob.php:
```diff
@@ -0,0 +1,61 @@
+
+    /**
+     * @param SecurityAlertCheckTask $checkTask
+     * @return SecurityAlertCheckJob
```
Prefer `return $this`, but this is also true

robbieaverill reviewed (approved):
Awesome!
